onmt/decoders/transformer.py

```diff
@@ -57,45 +82,49 @@ def forward(self, inputs, memory_bank, src_pad_mask, tgt_pad_mask,
             (FloatTensor, FloatTensor):
 
             * output ``(batch_size, 1, model_dim)``
-            * attn ``(batch_size, 1, src_len)``
+            * attns ``(batch_size, head, 1, src_len)``
 
         """
         dec_mask = None
+
         if step is None:
             tgt_len = tgt_pad_mask.size(-1)
-            future_mask = torch.ones(
-                [tgt_len, tgt_len],
-                device=tgt_pad_mask.device,
-                dtype=torch.uint8)
-            future_mask = future_mask.triu_(1).view(1, tgt_len, tgt_len)
-            # BoolTensor was introduced in pytorch 1.2
-            try:
-                future_mask = future_mask.bool()
-            except AttributeError:
-                pass
-            dec_mask = torch.gt(tgt_pad_mask + future_mask, 0)
+            if not future:  # apply future_mask, result mask in (B, T, T)
+                future_mask = torch.ones(
+                    [tgt_len, tgt_len],
+                    device=tgt_pad_mask.device,
+                    dtype=torch.uint8)
+                future_mask = future_mask.triu_(1).view(1, tgt_len, tgt_len)
+                # BoolTensor was introduced in pytorch 1.2
+                try:
+                    future_mask = future_mask.bool()
+                except AttributeError:
+                    pass
+                dec_mask = torch.gt(tgt_pad_mask + future_mask, 0)
+            else:  # only mask padding, result mask in (B, 1, T)
+                dec_mask = tgt_pad_mask
 
         input_norm = self.layer_norm_1(inputs)
 
         if isinstance(self.self_attn, MultiHeadedAttention):
-            query, attn = self.self_attn(input_norm, input_norm, input_norm,
-                                         mask=dec_mask,
-                                         layer_cache=layer_cache,
-                                         attn_type="self")
+            query, _ = self.self_attn(input_norm, input_norm, input_norm,
+                                      mask=dec_mask,
+                                      layer_cache=layer_cache,
+                                      attn_type="self")
         elif isinstance(self.self_attn, AverageAttention):
-            query, attn = self.self_attn(input_norm, mask=dec_mask,
-                                         layer_cache=layer_cache, step=step)
+            query, _ = self.self_attn(input_norm, mask=dec_mask,
+                                      layer_cache=layer_cache, step=step)
 
         query = self.drop(query) + inputs
 
         query_norm = self.layer_norm_2(query)
-        mid, attn = self.context_attn(memory_bank, memory_bank, query_norm,
-                                      mask=src_pad_mask,
-                                      layer_cache=layer_cache,
-                                      attn_type="context")
+        mid, attns = self.context_attn(memory_bank, memory_bank, query_norm,
```

Comment:
mid?

Reply:
I guess this naming is just because context attention block is at the middle of the decoder (self-attention -> context-attention -> feed forward) block.